### PR TITLE
Chore/abstract logging

### PR DESCRIPTION
**Pull request by Jakub-Vacek**
## Description

This PR adds centralized Logger to unify CLI outputs.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have updated the documentation accordingly. For updating Oclif commands documentation use [oclif-dev](https://github.com/oclif/dev-cli#oclif-dev-readme).
- [x] I have read the **CONTRIBUTION_GUIDE** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/common/command.abstract.ts [hunk 1]
```diff
@@ -10,4 +12,12 @@ export abstract class Command extends OclifCommand {
     }),
     help: flags.help({ char: 'h' }),
   };
+
+  /**
+   * Sets up logger for current run of oclif command
+   * @param quiet - quiet flag
+   */
+  public setUpLogger(quiet?: boolean) {
```
Comment (lukas-valenta) [inline src/common/command.abstract.ts:20]:
```suggestion
  protected void setUpLogger(quiet?: boolean) {
```

File: src/common/log.ts [hunk 2]
```diff
@@ -1,5 +1,56 @@
...
+export class Logger {
+  public static readonly successPrefix = '🆗';
+  public static readonly errorPrefix = '❌';
+  public static readonly warningPrefix = '⚠️';
+  private static logger: Logger | undefined = undefined;
+  public readonly quiet: boolean;
+
+  private constructor(quiet?: boolean) {
+    this.quiet = quiet || false;
+  }
```
Comment (lukas-valenta) [inline src/common/log.ts:14]:
```suggestion
  private constructor(private readonly quiet?: boolean) {}
```

**Review by lukas-valenta — COMMENTED** [review]
the overall approach looks good, much better than before. few things i'd change:
- rather than checking quiet in every log and then using process.stdout, i'd create different logger instances - StdoutLogger for normal, non quiet logging, DummyLogger for quiet logging and probably MockLogger that would just store logs in an array for tests
- i'd store all the strings in a single place as a record of either string or string-returning function (for substitutions). this would make it easier to change texts and we'd get rid of string duplication in tests

File: src/common/log.ts [hunk 3]
```diff
@@ -1,53 +1,107 @@
...
 
     return Logger.logger;
   }
 
-  public error(input: string): void {
-    const logger = Logger.getInstance();
-    if (!logger.quiet)
-      process.stderr.write(red(`${Logger.errorPrefix} ${input}`));
+  public static error(input: string): void {
+    Logger.getInstance().error(red(`${Logger.errorPrefix} ${input}`) + '\n');
```
Comment (lukas-valenta) [inline src/common/log.ts:92]: i'd put the formatting in StdoutLogger

File: src/common/log.ts [hunk 4]
```diff
@@ -1,53 +1,107 @@
...
   }
 
   public static info(input: string): void {
```
Comment (lukas-valenta) [inline src/common/log.ts:95]: for all of those i'd put it in ILogger interface

File: src/common/messages.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export const Messages = {
```
Comment (lukas-valenta) [inline src/common/messages.ts]: :+1:

**Comment by Jakub-Vacek** [comment]
@lukas-valenta  One more thing - should we make Logger take only `messages` inputs? That way we would enforce centralized messages.

**Comment by lukas-valenta** [comment]
@Jakub-Vacek that would be pretty cool :) can you pretty type it? e.g. `info('profile-created', 'some profile name')`. not sure if that'll work with the separation you did just after i looked at it, but you can just separate it outside the Messages object, `Messages = { ...common, ...install }` etc

**Comment by lukas-valenta** [comment]
wait you didn't push anything new, am i going insane? :thinking:

**Comment by Jakub-Vacek** [comment]
I didn't push anything. I have (unrelated to the design of logger) local changes (basically use of the logger in code and in tests) which I didn't want to push to keep this PR readable. If it is confusing I will push everything.

I will look into typing it :)

File: src/common/log.ts [hunk 6]
```diff
@@ -1,27 +1,154 @@
...
+    }
+
+    return Logger.logger;
+  }
+
+  private static getMessage<K extends keyof typeof messages>(
+    messsageTemplate: K,
+    ...args: Parameters<typeof messages[K]>
+  ): string {
+    return (messages[messsageTemplate] as (
```
Comment (Jakub-Vacek) [inline src/common/log.ts:122]: @lukas-valenta @TheEdward162 any idea how to get rid of cast? My goal is to "lock" logger to be usable only with messages. Fist argument id message template - key in messages and second are argument for that template - TS is smart enough to expect arguments matching to message template.

Comment (lukas-valenta) [reply src/common/log.ts:135]: i can't seem to get it done :disappointed:

Comment (TheEdward162) [reply src/common/log.ts:135]: This has to be an inference bug though, right? It should work otherwise.

File: src/common/log.ts [hunk 7]
```diff
@@ -1,27 +1,167 @@
...
+interface ILogger {
+  info(input: string): void;
+  warn(input: string): void;
+  success(input: string): void;
+  error(input: string): void;
+}
+class StdoutLogger implements ILogger {
+  public static readonly successPrefix = '🆗';
+  public static readonly errorPrefix = '❌';
+  public static readonly warningPrefix = '⚠️';
```
Comment (TheEdward162) [inline src/common/log.ts:15]: I would consider adding a `--no-emoji` flag or env var for users that don't want emoji in the output or don't have their terminal configured to correctly display it anyway.

File: src/common/log.ts (hunk 6 above)
Comment (lukas-valenta) [reply src/common/log.ts:135]: more limitation than bug

Comment (Jakub-Vacek) [reply src/common/log.ts:135]: Can I leave it as it is or should I look for some other solutions?

Comment (lukas-valenta) [reply src/common/log.ts:135]: leave it for now, i'm trying some stuff

**Review by martinalbert — COMMENTED** [review]
Looks great. Found just some minor things, mainly formatting or typos.

**Comment by Jakub-Vacek** [comment]
@martinalbert @lukas-valenta I will fix the typos/comments.

**Review by martinalbert — APPROVED** [review]

File: src/common/messages.ts [hunk 8]
```diff
@@ -6,40 +6,40 @@ const common = {
...
   initSuperJson: (path: string) => `Initialized "super.json" at path "${path}"`,
   updateSuperJson: (path: string) => `Updated "super.json" at path "${path}"`,
-  writeProfile: (path: string) => `Profile saved to "${path}"`,
+  writeProfile: (path: string) => `Profile saved to: "${path}"`,
   unableToWriteProfile: (profile: string, err: unknown) =>
     // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
-    `Could not write profile ${profile} source: ${err}`,
-  invalidProfileId: (id: string) => `Invalid profile id "${id}"`,
-  invalidProfileName: (name: string) => `Invalid profile name "${name}"`,
+    `Could not write profile: "${profile}"" source: "${err}"`,
```
Comment (martinalbert) [inline src/common/messages.ts:12]: double quotes after `${profile}`

File: src/common/messages.ts [hunk 9]
```diff
@@ -6,40 +6,40 @@ const common = {
...
+    `Superface server responded with error: "${name}": "${message}"`,
 };
 
 const load = {
   localProfileFound: (profileId: string, path: string) =>
-    `Profile "${profileId}" found on local file system at path "${path}"`,
+    `Profile "${profileId}" found on local file system at path: "${path}"`,
   localMapFound: (profile: string, provider: string, path: string) =>
-    `Map for profile "${profile}" and provider "${provider}" found on local file system at path "${path}"`,
+    `Map for profile: "${profile}" and provide:r "${provider}" found on local file system at path: "${path}"`,
```
Comment (martinalbert) [inline src/common/messages.ts:31]: typo in `provider`

**Review by martinalbert — APPROVED** [review]

**Pull request merged**